crates/fuel-core/src/service/adapters/consensus_parameters_provider.rs

```diff
@@ -120,23 +170,29 @@ impl RunnableTask for Task {
             }
 
             Some(event) = self.blocks_events.next() => {
-                let new_version = event
+                let header = event
                     .sealed_block
                     .entity
-                    .header()
-                    .consensus_parameters_version();
-
-                if new_version > *self.shared_state.latest_consensus_parameters_version.lock() {
-                    match self.shared_state.cache_consensus_parameters(new_version) {
-                        Ok(_) => {
-                            *self.shared_state.latest_consensus_parameters_version.lock() = new_version;
-                        }
-                        Err(err) => {
-                            tracing::error!("Failed to cache consensus parameters: {:?}", err);
-                            return TaskNextAction::Stop
-                        }
-                    }
+                    .header();
+
+                let new_consensus_parameters_version = header.consensus_parameters_version();
+                if !update_cached_value(
+                    &self.shared_state.latest_consensus_parameters_version,
+                    new_consensus_parameters_version,
+                    || self.shared_state.cache_consensus_parameters(new_consensus_parameters_version).map(|_| ())
+                ) {
```

Comment:
You have a deadlock here.  `self.shared_state.cache_consensus_parameters` is doing write lock inside and `update_cached_value` is doing upgradable lock